kernel/src/hil/time.rs

```diff
@@ -1,33 +1,270 @@
-//! Hardware agnostic interfaces for counter-like resources.
+//! Hardware agnostic interfaces for time and timers within the Tock
+//! kernel. These traits are designed to be able encompass the wide
+//! variety of hardare counters in a general yet efficient way. They
+//! abstract the frequency of a counter through the `Frequency` trait
+//! and the width of a time value through the `Ticks`
+//! trait. Higher-level software abstractions should generally rely on
+//! standard and common implementations of these traits (e.g.. `u32`
+//! ticks and 16MHz frequency).  Hardware counter implementations and
+//! peripherals can represent the actual hardware units an translate
+//! into these more general ones.
 
 use crate::ReturnCode;
+use core::cmp::{Eq, Ord, Ordering, PartialOrd};
+use core::fmt;
 
-pub trait Time<W = u32> {
+/// An integer type defining the width of a time value, which allows
+/// clients to know when wraparound will occur.
+
+pub trait Ticks: Clone + Copy + From<u32> + fmt::Debug + Ord + PartialOrd + Eq {
```

Comment:
Given that we now support 64 bit timers below the HIL (RISC-V), should this also implement `From<u64>`?

Reply:
It actually shouldn't. You can't make a Ticks32 from a 64 bit value. The presence of From<u32> is basically for the syscall driver, so it can convert from 32 bit values into Ticks. The assumption is that it is on top of a Ticks that is >= 32 bits in width. The nRF 24 bit Ticks have to expand to 32 (e.g.. by handing overflow) for the syscall driver to use.

Reply:
Note https://github.com/tock/tock/blob/a97b7cd66b6ed7e8d3e5658d51f0155de8c82c1d/kernel/src/hil/time.rs#L433